### PR TITLE
Rearm TimerActions

### PR DESCRIPTION
**Pull request by glommer**
### What does this PR do?

Provides a way for us to rearm TimerAction
This is a common primitive found in many timer libraries.
As we prepare for that, it makes sense to split TimerAction in two, since it doesn't make sense to rearm a repeatable TimerAction

### Checklist

[x] I have added unit tests to the code I am submitting
[x] My unit tests cover both failure and success scenarios
[] If applicable, I have discussed my architecture
[x] The new code I am adding is formatted using `rustfmt`

### CODE REVIEW

File: scipio/src/timer.rs [hunk 1]
```diff
@@ -6,18 +6,47 @@
...
+            Reactor::get().remove_timer(self.id);
+        }
+
+        // Update the timeout.
+        self.when = Instant::now() + dur;
+
+        if let Some(waker) = self.waker.as_mut() {
+            // Re-register the timer with the new timeout.
+            Reactor::get().insert_timer(self.id, self.when, waker);
+        }
```
Comment (HippoBaro) [inline scipio/src/timer.rs:39]: Can you explain why we update `self.when` even if we don't insert a new timer? Also, in what circunstances `self.waker` can be `None`?

Comment (glommer) [reply scipio/src/timer.rs:39]: I can, but notice please that this code is pre-existing, I am just moving here.

`self.waker` starts as None. It only gets assigned when the future is first polled() (call to `await`).

Imagine for instance that you schedule a function that will call `timer.await` for execution in a task queue that is very busy: it may take a long time for `waker` not to be `None`. There are no guarantees as to when that might happen.

We only need to go to the reactor on reset if the future was already polled and is pending (has a waker).
If the future was not polled, then all we have to do is update `when`: whenever it does get polled, it will be polled with the right `when`

In summary:
`when` has to be updated in all cases: 

* if the future was polled already, so we always have the most up2date information (waker will be `Some`, because we are waiting on it)
 * if the future was not polled, so we register the timer correctly for the right expiration (waker will be `None`)

File: scipio/src/timer.rs [hunk 2]
```diff
@@ -283,48 +300,175 @@ impl<T: 'static> TimerAction<T> {
...
+    /// }).unwrap();
+    /// handle.join().unwrap();
+    /// ```
+    /// [`TimerActionOnce`]: struct.TimerActionOnce
+    /// [`Option`]: https://doc.rust-lang.org/std/option/enum.Option.html
+    pub async fn join(self) -> Option<T> {
+        self.handle.await
+    }
+
+    /// Rearm a [`TimerActionOnce`], so it fires at the specified [`Duration`] from now
```
Comment (HippoBaro) [inline scipio/src/timer.rs:416]: "at the specified [`Duration`] from now" -> "in the specified [`Duration`] from now"

File: scipio/src/timer.rs [hunk 3]
```diff
@@ -283,48 +300,175 @@ impl<T: 'static> TimerAction<T> {
...
+    /// handle.join().unwrap();
+    /// ```
+    /// [`TimerActionOnce`]: struct.TimerActionOnce
+    /// [`Duration`]: https://doc.rust-lang.org/std/time/struct.Duration.html
+    pub fn rearm_in(&self, dur: Duration) {
+        let mut inner = self.inner.borrow_mut();
+        inner.reset(dur);
+    }
+
+    /// Rearm a [`TimerActionOnce`], for a specified [`Instant`]
```
Comment (HippoBaro) [inline scipio/src/timer.rs:439]: "/// Rearm a [`TimerActionOnce`], so it fires at the specified [`Instant`]"

For consistency with the other variant.

**Review by HippoBaro — APPROVED** [review]
Some nits, otherwise all good :+1:
